modules/core/src/test/java/org/apache/ignite/internal/client/thin/ServiceAwarenessTest.java

```diff
@@ -0,0 +1,624 @@
+
+        // Service topology on the client.
+        Set<UUID> srvcTopOnClient = new GridConcurrentHashSet<>();
+
+        addSrvcTopUpdateClientLogLsnr(srvcTopOnClient::addAll);
+
+        AtomicBoolean svcRunFlag = new AtomicBoolean(true);
+
+        try (IgniteClient client = startClient()) {
+            ServicesTest.TestServiceInterface svc = client.services().serviceProxy(SRV_NAME, ServicesTest.TestServiceInterface.class);
+
+            runAsync(() -> {
+                while (svcRunFlag.get())
+                    svc.testMethod();
+            });
+
+            waitForCondition(() -> srvcTopOnClient.size() == INIT_SRVC_NODES_CNT
+                && srvcTopOnClient.contains(grid(1).localNode().id())
+                && srvcTopOnClient.contains(grid(2).localNode().id()),
+                getTestTimeout());
+
+            // Delays service redeployment and the service topology update on the server side.
+            testDisco.toBlock.add(ServiceClusterDeploymentResultBatch.class);
+
+            startGrid(GRIDS);
+
+            waitForCondition(() -> testDisco.blocked.size() == 1, getTestTimeout());
+
+            // Ensure all the nodes have started but the service topology hasn't updated yet.
+            for (Ignite ig : G.allGrids()) {
+                assertEquals(ig.cluster().nodes().size(), GRIDS + 1);
+
+                // Ensure there are still SRVC_FILTERED_NOIDES_CNT nodes with the service instance.
+                assertEquals(((IgniteEx)ig).context().service().serviceTopology(SRV_NAME, 0).size(),
+                    INIT_SRVC_NODES_CNT);
+            }
+
+            // Ensure the client's topology is not updated.
+            assertTrue(srvcTopOnClient.size() == INIT_SRVC_NODES_CNT
+                && !srvcTopOnClient.contains(grid(GRIDS).localNode().id()));
+
+            testDisco.release();
+
+            // Ensure the service topology has been updated to 3 instances per cluster.
+            for (Ignite ig : G.allGrids()) {
+                waitForCondition(
+                    () -> {
+                        try {
+                            return ((IgniteEx)ig).context().service().serviceTopology(SRV_NAME, 0).size() == 3;
+                        }
+                        catch (Exception e) {
+                            return false;
+                        }
+                    },
+                    getTestTimeout()
+                );
+            }
+
+            waitForCondition(() -> srvcTopOnClient.size() == 3 && srvcTopOnClient.contains(grid(1).localNode().id())
+                && srvcTopOnClient.contains(grid(2).localNode().id())
+                && srvcTopOnClient.contains(grid(GRIDS).localNode().id()), getTestTimeout());
+        }
+        finally {
+            svcRunFlag.set(false);
+        }
+    }
+
+    /**
+     * Tests several nodes come while one thread is used to call the service.
+     */
+    @Test
+    public void testNodesComeOneThread() throws Exception {
```

Comment:
```suggestion
    public void testNodesJoinSingleThreaded() throws Exception {
```

Let's use common terminology - here and below.